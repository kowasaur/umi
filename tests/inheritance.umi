class Animal {
    String name

    Animal(String name) { this.name = name }

    Void speak() { print("Default sound") }

    String toString() { "animal named " + name }
}

class Cat : Animal {
    Cat() {}

    Void speak() { meow() }

    Void meow() { print("Meow") }
}

class Dog : Animal {
    Int age

    Dog(String name, Int age) {
        base(name)
        this.age = age
    }
}

class Human : Animal {
    Human(String name) { base(name) }

    Void speak() { 
        print("Hello my name is " + name)
        print("I am not an " + toString())
    }
}

Void main() {
    Animal a = Animal("alfred")
    a.speak()
    print(a.toString())

    Animal cat = Cat()
    cat.speak()
    (cat as Cat).meow()

    Dog dog = Dog("spot", 5)
    print(dog.toString())
    print(dog.age)

    Animal joe = Human("Joe")
    joe.speak()
}
