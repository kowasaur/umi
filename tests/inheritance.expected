Default sound
animal named alfred
Meow
Meow
majestic animal named Mittens
animal named spot
5
Default sound
Hello my name is Joe
I am not an animal named Joe
